Add similarity topic (#11533)
opensearch-project/documentation-website: _im-plugin/similarity.md

---
layout: default
title: Similarity
nav_order: 37
---

# Similarity

A similarity defines how matching documents are scored and ranked during search operations. OpenSearch uses similarity algorithms to calculate relevance scores that determine the order of search results. 

> Each field can have its own similarity configuration, allowing fine-tuned control over how different types of content are scored. You can define custom similarity algorithms in your index settings at index level. Once configured, you can apply these algorithms to specific fields using the [`similarity` mapping parameter]({{site.url}}{{site.baseurl}}/mappings/mapping-parameters/similarity/).

Reviewer: ```suggestion
Each field can have its own similarity configuration, allowing fine-tuned control over how different types of content are scored. You can define custom similarity algorithms in your index settings at the index level. Once configured, you can apply these algorithms to specific fields using the [`similarity` mapping parameter]({{site.url}}{{site.baseurl}}/mappings/mapping-parameters/similarity/).
```

You can set the default similarity when creating an index:

```json
PUT /product_catalog
{
  "settings": {
    "index": {
      "similarity": {
        "default": {
          "type": "boolean"
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

When you use the special `default` name, the similarity is automatically applied to all fields in the index. You don't need to specify the similarity in individual field mappings.

### Changing default similarity after index creation

To change the default similarity after index creation, you must close the index, update the settings, and reopen it:

```json
POST /product_catalog/_close
```
{% include copy-curl.html %}

```json
PUT /product_catalog/_settings
{
  "index": {
    "similarity": {
      "default": {
        "type": "DFR",
        "basic_model": "g",
        "after_effect": "l",
        "normalization": "h2",
        "normalization.h2.c": "3.0"
      }
    }
  }
}
```
{% include copy-curl.html %}

```json
POST /product_catalog/_open
```
{% include copy-curl.html %}

To verify that the default similarity has changed, send the following request:

```json
GET /product_catalog/_settings
```
{% include copy-curl.html %}

> The response confirms that the default similarity is updated:

Reviewer: ```suggestion
The response confirms that the default similarity has been updated:
```

        "basic_model": null,
        "after_effect": null,
        "normalization": null,
        "normalization.h2.c": null
      }
    }
  }
}
```
{% include copy-curl.html %}

## Configuring a built-in similarity on the field level

OpenSearch provides built-in similarity algorithms (`BM25` and `boolean`) that can be used directly in field mappings. To configure similarity on the field level, apply the similarity to the field when creating the index. The following example applies a `boolean` similarity to the `content` field:

```json
PUT /blog_posts_2/
{
  "mappings": {
    "properties": {
      "content": {
        "type": "text",
        "similarity": "boolean"
      }
    }
  }
}
```
{% include copy-curl.html %}

In this example, only the `content` field uses boolean similarity. Other fields in the index will use the default BM25 similarity.

## Configuring a custom similarity

For more advanced use cases, you can define named custom similarities that can be applied to specific fields. This approach provides greater flexibility when different fields in your index need different scoring algorithms.

Configuring named custom similarities requires a two-step process:

> 1. Define the similarity in index settings with a custom name

Reviewer: ```suggestion
1. Define the similarity in the index settings with a custom name.
```

> 2. Apply the similarity to specific fields in mappings

Reviewer: ```suggestion
2. Apply the similarity to specific fields in mappings.
```

### Step 1: Define a custom similarity

The following example creates an index with a custom DFR similarity configuration:

```json
PUT /blog_posts
{
  "settings": {
    "index": {
      "similarity": {
        "blog_similarity": {
          "type": "DFR",
          "basic_model": "g",
          "after_effect": "l",
          "normalization": "h2",
          "normalization.h2.c": "3.0"
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

### Step 2: Apply the custom similarity to fields

After defining the similarity, you must explicitly apply it to specific fields in your mappings. Unlike default similarity, custom similarities are not automatically applied to all fields:

```json
PUT /blog_posts/_mapping
{
  "properties": {
    "content": {
      "type": "text",
      "similarity": "blog_similarity"
    }
  }
}
```
{% include copy-curl.html %}

> In this example, only the `content` field uses the custom `blog_similarity`. Other fields in the index (if any) continue using the default BM25 similarity unless explicitly configured otherwise.

Reviewer: ```suggestion
In this example, only the `content` field uses the custom `blog_similarity`. Other fields in the index (if any) continue using the default `BM25` similarity unless explicitly configured otherwise.
```

## Available similarity types

OpenSearch supports the following similarity types.

### BM25 similarity (default)

`BM25` similarity is a TF/IDF-based similarity with built-in term frequency normalization. It works well for most text fields, particularly shorter fields like titles and names.

BM25 similarity supports the following parameters.

| Parameter | Description | Default | Required |
|-----------|-------------|---------|----------|
| `k1` | Controls non-linear term frequency normalization (saturation). | `1.2` | No |
| `b` | Controls how much document length normalizes term frequency values. | `0.75` | No |
| `discount_overlaps` | Determines whether overlap tokens (tokens with 0 position increment) are ignored when computing norms. | `true` (ignore overlap tokens when computing norms) | No |

### Boolean similarity

`boolean` similarity is a built-in similarity that assigns all matching documents the same constant score, making it useful when you only care about whether documents match rather than how relevant they are. This similarity ignores term frequency, document length, and other scoring factors.

Boolean similarity does not support parameters.

### DFR similarity

> `DFR` similarity implements the [divergence from randomness](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/DFRSimilarity.html) framework for document scoring.

Reviewer: Link gives 404 Not Found

`DFR` similarity supports the following parameters.

| Parameter | Description | Valid values | Required |
|-----------|-------------|------------------|----------|

> | `basic_model` | Basic model for the DFR framework. | [`g`](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/BasicModelG.html), [`if`](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/BasicModelIF.html), [`in`](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/BasicModelIn.html), [`ine`](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/BasicModelIne.html) | Yes |

Author: That's because there's a variable in the links that is resolved when the site is built.

### DFI similarity

`DFI` similarity implements the divergence from independence model.

`DFI` similarity supports the following parameters.

| Parameter | Description | Valid values | Required |
|-----------|-------------|------------------|----------|
| `independence_measure` | Independence measure for the DFI model. | [`standardized`](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/IndependenceStandardized.html), [`saturated`](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/IndependenceSaturated.html), [`chisquared`](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/IndependenceChiSquared.html) | Yes |

When using DFI similarity, avoid removing stop words for optimal relevance. Terms with frequency less than expected will receive a score of 0.

### IB similarity

> `IB` similarity uses the [information-based model](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/IBSimilarity.html) that analyzes the repetitive usage of basic elements in symbolic distributions.

Reviewer: ```suggestion
`IB` similarity uses the [information-based model](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/IBSimilarity.html), which analyzes the repetitive usage of basic elements in symbolic distributions.
```

`IB` similarity supports the following parameters.

| Parameter | Description | Valid values | Required |
|-----------|-------------|------------------|----------|
| `distribution` | Distribution model for the IB framework. | [`ll`](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/DistributionLL.html), [`spl`](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/DistributionSPL.html) | Yes |
| `lambda` | Lambda model for the IB framework. | [`df`](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/LambdaDF.html), [`ttf`](https://lucene.apache.org/core/{{site.lucene_version}}/core/org/apache/lucene/search/similarities/LambdaTTF.html) | Yes |

> | `normalization` | Normalization model for the IB framework. | Same options as DFR similarity | Yes |

Reviewer: ```suggestion
| `normalization` | Normalization model for the IB framework. | Same options as `DFR` similarity | Yes |
```

                  "value": 9,
                  "description": "field.sumDocFreq",
                  "details": []
                },
                {
                  "value": 9,
                  "description": "field.sumTotalTermFreq",
                  "details": []
                },
                {
                  "value": 1,
                  "description": "term.docFreq",
                  "details": []
                },
                {
                  "value": 1,
                  "description": "term.totalTermFreq",
                  "details": []
                },
                {
                  "value": 1,
                  "description": "doc.freq",
                  "details": []
                },
                {
                  "value": 5,
                  "description": "doc.length",
                  "details": []
                }
              ]
            }
          ]
        }
      }
    ]
  }
}
```

In this optimized example:

- The `weight_script` calculates the score of `2.8109303` (IDF + query boost, same for all documents).
- The main `script` calculates document-specific factors using the pre-calculated `weight` value.

> - The final score is `1.2570862` (identical to the single-script approach, but more efficient).

Reviewer: ```suggestion
- The final score is `1.2570862` (identical to the single-script approach but more efficient).
```
